### ISSUE TITLE
Multiple terraform_remote_state giving the same result

### ISSUE THREAD

**Issue by keeperpl** [hashicorp/terraform#17316]
### Terraform Version
```
Terraform v0.11.3
+ provider.azurerm v1.1.1
+ provider.random v1.1.0
+ provider.template v1.0.0
```

### Terraform Configuration Files

I'm upgrading terraform from 0.10.8 and terraform provider 1.0.1 to terraform 0.11.3.
We have 3 environments with each having a remote terraform state file in azure blob.

DEV Environment is using terraform_remote_state to get vnet_ids of other environments to configure peering to both remote vnets.
```hcl
data "terraform_remote_state" "staging" {
    backend = "azure"
    config {
        storage_account_name = "tfstate"
        container_name       = "tfstate-files"
        key                  = "staging.tfstate"
        access_key           = ""
    }
}

data "terraform_remote_state" "prod" {
    backend = "azure"
    config {
        storage_account_name = "tfstate"
        container_name       = "tfstate-files"
        key                  = "prod.tfstate"
        access_key           = ""
    }
}
```

Both staging and prod environments are using outputs to store proper vnet ids in remote terraform state file:

```hcl
output "vnet_id" {
  value = "${azurerm_virtual_network.vnet.id}"
}
```

Staging state file fragment:

```
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {
                "vnet_id": {
                    "sensitive": false,
                    "type": "string",
                    "value": "/subscriptions/..../resourceGroups/staging/providers/Microsoft.Network/virtualNetworks/staging-net"
                },
```

Production state file fragment:
```
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {
                "vnet_id": {
                    "sensitive": false,
                    "type": "string",
                    "value": "/subscriptions/..../resourceGroups/prod/providers/Microsoft.Network/virtualNetworks/prod-net"
                },
```
Despite that, using outputs in dev environment:

```hcl
output "staging vnet_id" {
  value = "${data.terraform_remote_state.staging.vnet_id}"
}

output "prod vnet_id" {
  value = "${data.terraform_remote_state.prod.vnet_id}"
}
```

are showing one, the same vnet_id twice. Always the one from environment that had remote terraform state refreshed as last.

### Expected Behavior

Output from terraform_remote_states should be different for different remote states.

### Actual Behavior
Output from different terraform_remote_states is the same.

### Additional Context
The same behaviour can be reproduced in my case when I leave terraform at 0.10.8 and upgrade terraform provider to 1.0.2.
With TF 0.10.8 and TF provider 1.0.1 everything is working correctly.

I tried renaming "vnet_id" to "vnet_id_prod". It didn't help. Running terraform apply in staging makes vnet_id_prod not accessible for dev. Running terraform apply in prod sets all vnet_ids accessible for dev as /subscriptions/..../resourceGroups/prod/providers/Microsoft.Network/virtualNetworks/prod-net

**Comment by jbardin** [comment]
Hi @keeperpl,

Thanks for filing the issue.

It appears that this is caused by the deprecation wrapper for the backend not taking into account the possibility of multiple remote state configurations.

While we still need to fix the issue internally, you should be able to work around the problem by using the correct backend name of `azurerm`.

**Comment by apparentlymart** [comment]
Hi again @keeperpl! Sorry for the long silence here.

This was actually fixed in fe527ec9d7f1f4a1dd27942ed7f695416736e57b as part of some other work, but since we weren't intentionally trying to address this issue we neglected to post an update here. That work is now on master and ready to be included in the forthcoming v0.12.0 release.

As James said, the problem here was that the legacy name `"azure"` was referring to a single backend instance rather than creating one instance per usage as with other backends. This was an oversight in how the deprecation of that name was implemented.

We still recommend moving to the name `"azurerm"` for this backend since the old name is likely to be removed in a future release.
